lib/Parse/ParseDecl.cpp

```diff
@@ -2906,27 +2906,43 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
     }
     
     // Parse the subject.
-    if (Tok.isContextualKeyword("set")) {
-      consumeToken();
-    } else {
-      diagnose(Loc, diag::attr_access_expected_set, AttrName);
+    if (!Tok.isContextualKeyword("set")) {
+      auto diag = diagnose(Loc, diag::attr_access_expected_set, AttrName);
+      
+      if (Tok.is(tok::l_paren)) {
+        return makeParserSuccess();
+      }
 
       // Minimal recovery: if there's a single token and then an r_paren,
       // consume them both. If there's just an r_paren, consume that.
-      if (!consumeIf(tok::r_paren)) {
-        if (Tok.isNot(tok::l_paren) && peekToken().is(tok::r_paren)) {
-          consumeToken();
-          consumeToken(tok::r_paren);
-        }
+      if (Tok.is(tok::r_paren)) {
+        auto SetLoc = consumeToken(tok::r_paren);
+        diag.fixItInsert(SetLoc, "set");
+      } else if (peekToken().is(tok::r_paren)) {
+        auto SetLoc = consumeToken();
+        diag.fixItReplace(SetLoc, "set");
+        consumeToken(tok::r_paren);
+      } else if (isStartOfSwiftDecl()) {
+        diag.fixItInsertAfter(LParenLoc, "set)");
+      } else {
+        BacktrackingScope Scope(*this);
+        auto InvalidTokLoc = consumeToken();
+        
+        if (isStartOfSwiftDecl())
```

Comment:
I'm using `isStartOfSwiftDecl` here to check if a declaration follows the missing `)` (optionally with an unexpected token) for the following cases:

- `private( <declaration>` inserts `set)` to be `private(set) <declaration>`
- `private(<unexpected> <declaration>` replaces `<unexpected>` with `set)` to be `private(set) <declaration>`

I'm not sure if it'd add an overhead to use it only for fix-its though. We can remove its use and not supported these two cases (or maybe replace it with a lighter check like checking for the `var` keyword).

I'd love to hear your thoughts.

Reply:
@lucianopa-msft I have a quick question since you mentioned it's just a failure path, do we need the `BacktrackingScope` here or is it okay to consume the invalid tokens and not revert back at all?

Reply:
Makes sense, thanks for the clarification 🙏🏼